### PR TITLE
Apply signed integer settings sent in the legacy binary format, and validate non-zero settings there

### PR DESCRIPTION
**Pull request by groeneai**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):

Fixed two defects in the legacy binary settings format, used for native-protocol clients declaring a revision below 54429. Signed integer settings (for example `max_partitions_to_read` or `iceberg_snapshot_id`) were silently ignored, and were reported as changed while holding their default, so they overrode the table-level value with that default. Settings required to be greater than zero (for example `max_block_size`) were not validated when read from that format.

### Description

Two defects in the legacy binary settings format, which `TCPHandler` selects for any native-protocol peer declaring a revision below `DBMS_MIN_REVISION_WITH_SETTINGS_SERIALIZED_AS_STRINGS` (54429); no minimum revision is enforced.

**1. Signed integral settings were discarded.** `SettingFieldNumber<T>::readBinary` decoded into a local, then assigned from `value`, its own data member. `BaseSettings::read` resets to defaults first and `operator=` sets `changed`, so the setting was flagged as a user override while holding its default, and `ReadFromMergeTree` then prefers that default over the table-level value. 22 signed query settings are affected, six of them snapshot selectors for Iceberg, DeltaLake and Paimon. Introduced in `300727afa316a01c` (2020-07-27).

**2. A non-zero setting was not validated.** `SettingFieldNonZeroUInt64` re-runs `checkValueNonZero()` from every mutating entry point it overrides, but not `readBinary`, whose base `*this = static_cast<T>(x)` binds statically to `SettingFieldUInt64::operator=`. Of the twelve operations in `SettingFieldOps`, `read_binary` was the only mutating one that could store a forbidden value. Requested by @ Algunenano in review.

Both writers were already correct, so the wire format is unchanged.

On (2), a TCP query's outcome is unchanged: `SettingsConstraints::checkOrClamp` already rejects the zero with the same `BAD_ARGUMENTS`, now raised at the decode step. The override matters where nothing runs after the decode: `QueryPlanSerializationSettings::readBinary` and the old-format branch of `DistributedAsyncInsertHeader`. The modern header format already rejects zero there, so this makes the legacy branch consistent; the one behaviour change is that a pre-2024-03 old-format batch carrying an explicit `0` would stop draining.

Tests. A raw native-protocol client pinned to revision 54428 requires `TOO_MANY_PARTITIONS` for `max_partitions_to_read = 1`. A gtest drives `Settings::read` with `SettingsWriteFormat::BINARY`, because the constraints check masks (2) on every SQL-reachable path. Both fail without this change.

---
Workflow [[PR](https://s3.amazonaws.com/clickhouse-test-reports/praktika.html?PR=118610&sha=latest&name_0=PR)]
Sync PR [[sync-upstream/pr/118610](https://github.com/search?q=head%3Async-upstream%2Fpr%2F118610+org%3AClickHouse+type%3Apr&type=pullrequests)]

### CODE REVIEW

**Comment by groeneai** [comment]
<details>
<summary><b>Internal second-model review</b> (click to expand)</summary>

Two independent passes before this PR was opened: an automated adversarial re-review by a
different model, and my own cold read of the resulting code (done before opening the
implementation notes, so the two passes are not anchored on each other).

**Automated pass: 0 findings.** Coverage: all changed files, the full numeric-setting
implementations and wrappers, serialization callers, the legacy TCP handshake and query paths,
MergeTree limit enforcement, and the existing related tests.

**My own pass: 6 findings, no blocker.** Two were in the PR description's prose and are
corrected in the description above. Four are nits, listed below with verdicts.

| | Finding | Verdict |
|---|---|---|
| ⚠️ | The description claimed 23 affected signed query settings. The count at this base is **22**: 20 `DECLARE(Int64, ...)` plus 2 `Int32` (`os_threads_nice_value_query`, declared via `DECLARE_WITH_ALIAS`, and `os_threads_nice_value_materialized_view`). Two further signed settings exist only as `MAKE_OBSOLETE` (`annoy_index_search_k_nodes`, `max_limit_for_ann_queries`), and those are accepted-and-ignored rather than affected. | **Agreed, corrected** in the description |
| ⚠️ | The description called four of them Iceberg/DeltaLake/Paimon time-travel selectors. There are **six** read-side snapshot or version selectors: `iceberg_snapshot_id`, `iceberg_timestamp_ms`, `paimon_target_snapshot_id`, and `delta_lake_snapshot_version` / `_start_version` / `_end_version`. The three `iceberg_expire_default_*` settings are expiry-policy defaults consumed by `expire_snapshots`, not selectors, which is where the undercount came from. | **Agreed, corrected** in the description |
| 💡 | The commit message still carries those two original figures. I did not amend it; the corrected numbers are the ones in the description above, which is what feeds the changelog. Flagging it so a reader of the commit list is not misled. | Noted, not blocking |
| 💡 | No in-range boundary arm (`max_partitions_to_read = 2`, expecting a data packet) beside the out-of-range assertion. I considered adding one and decided against it. The always-true mutant that pattern exists to catch is "throws `TOO_MANY_PARTITIONS` regardless of the setting", and the existing no-settings control already reddens on that, using the identical connection helper, handshake and query but requiring the opposite observable. The out-of-range arm on its own also already pins the transported value to exactly 1, since `ReadFromMergeTree` raises only when `max > 0` and `partitions.size()` (2) `> max`. The only mutant an in-range row would uniquely detect is changing that comparison to `>=`, which is a different subsystem's contract. The negative half of the signed range is verified end to end out of band (`-1` and `-5` both decode and correctly impose no limit), so nothing about the zigzag decode is unproven. If you would rather have the boundary row committed, it is a one-line addition plus one reference line and I am happy to push it. | Declined, with the boundary value already measured |
| 💡 | The test's `.python` falls back to port 9000 when `CLICKHOUSE_PORT_TCP` is unset, whereas the closest siblings use an invalid sentinel so a missing variable fails loudly. `shell_config.sh` always exports the real value, so this has no effect under `clickhouse-test`. | Noted, not blocking |
| 💡 | Local test-run leftovers (`*.stderr` / `*.stdout`) in the test directory. Gitignored and absent from the diff. | Harness artifact, no action |

**Checked and clean.** The fix is symmetric with the writer (`writeVarInt` out, `readVarInt` in),
so the wire format is untouched. I enumerated the invariant carriers independently before reading
the implementation notes: all 15 `SettingField*::readBinary` bodies assign from their own local or
a returned value, so the signed arm was the only carrier with the slip and no sibling fix is due.
I also proved the test's liveness by reading rather than relying on the recorded run:
`BaseSettings::read` calls `resetToDefault()`, so `value` is the default `-1` at that point, and
`operator=` sets `changed`, so `checkLimits` prefers `-1` and raises nothing, which is exactly the
`got packet 1` failure the base binary produced. The exception is raised during pipeline building,
before any header block is sent, so the test needs no block parsing; and no progress packet can
precede the header, because the header is sent and flushed before the executor loop that emits
progress. `max_partitions_to_read` appears in no CI test config and is not randomized, and parts
in different partitions never merge, so the two-partition precondition cannot be raced away.

Session id: cron:clickhouse-review-slot-8:20260907-153100

</details>

**Comment by groeneai** [comment]
<details>
<summary><b>Pre-PR validation gate</b> (click to expand)</summary>

| # | Question | Answer |
|---|----------|--------|
| a | **Deterministic repro?** | Yes. `tests/clickhouse-test 05136_legacy_binary_settings_signed_integer` fails on every run against a master binary and passes on every run with this change. Not probabilistic. |
| b | **Root cause explained?** | Yes. The signed arm of `SettingFieldNumber<T>::readBinary` decoded the wire value into a local and then assigned from `value`, its own data member. `BaseSettings::read` resets to defaults first and `operator=` sets `changed = true`, so the setting was flagged as a user override while holding its default, and `ReadFromMergeTree` then preferred that default over the table-level value. |
| c | **Fix matches root cause?** | Yes. One token at the defect site, assigning from the decoded local. No widened bound, no test tag, no defensive guard, no symptom-site check. |
| d | **Test intent preserved / new tests added?** | New stateless regression test (`.sh` + `.python` + `.reference`). No existing test was weakened, relaxed or retagged. |
| e | **Both directions demonstrated?** | Yes. Base binary (Build-ID `3159e7ac...`): FAIL, `expected an Exception packet, got packet 1`. Fixed binary (Build-ID `4a5834ea...`): OK. A third arm with the setting removed from the scenario fails the same assertion, proving the exception comes from the transported setting and not from the hand-rolled handshake. Server identity was asserted with `SELECT lower(buildId())` before each verdict. |
| f | **Fix is general across code paths?** | Yes. I read every `readBinary` body in `src/Core/SettingsFields.{h,cpp}`: `MaxThreads`, `Timespan`, `String`, `Map`, `Char`, `URI`, `EnumHelpers`, `Timezone`, `Custom`, `Enum`, `MultiEnum`, and the unsigned and floating-point arms of `SettingFieldNumber<T>` all assign from their own local. The signed arm was the only carrier with the slip. `SettingAutoWrapper` and `BaseSettings::readBinary` delegate to the same leaf and are fixed transitively. |
| g | **Fix generalizes across inputs (params/datatypes/wrappers)?** | Yes. `SettingFieldNumber<Int64>` and `<Int32>` share the single `if constexpr` branch, so both instantiations are fixed by the one token; `bool` and the unsigned types take the unsigned arm and `float`/`double` the string arm, both already correct. Values were exercised at 1 (out of range), 2 (the exact boundary), 3, -1 and -5, plus the no-setting case, and the transported value is honoured exactly in each. Type wrappers such as `Nullable` or `LowCardinality` do not apply: a setting field is a scalar, not a column. |
| h | **Backward compatible?** | Yes, and nothing is due. The symmetric writer is unchanged and was already correct, so the wire format does not change; no setting default or type changes, hence no `SettingsChangesHistory.cpp` entry and no gate. The only behavioural change is that a pre-54429 peer's signed settings are now applied instead of silently dropped, which is the fix. |
| i | **Invariants and contracts preserved?** | Yes. The invariant is that every `SettingField*::readBinary` assigns the value it just decoded rather than re-reading its own member; it now holds for all carriers. Byte consumption is unchanged, since `readVarInt` already consumed the value, so the settings stream stays in sync and every later setting in the packet parses exactly as before. `changed` is still set by `operator=` on the same single path, and no error or early-return path is added or altered. |

</details>

Session id: cron:clickhouse-impl-slot-4:20260907-142900

**Comment by groeneai** [comment]
cc @Algunenano @tavplubix, could you review this? The signed-integral arm of `SettingFieldNumber<T>::readBinary` assigned from its own `value` member instead of the local it had just decoded, so a signed setting sent in the legacy binary settings format (any native-protocol peer declaring a revision below 54429) was discarded and left flagged as changed while holding its default, which then beat the table-level value in `ReadFromMergeTree`.

File: src/Core/SettingsFields.cpp [hunk 1]
```diff
@@ -260,7 +260,7 @@ void SettingFieldNumber<T>::readBinary(ReadBuffer & in)
     {
         Int64 x = 0;
         readVarInt(x, in);
-        *this = static_cast<T>(value);
+        *this = static_cast<T>(x);
```
[the remark this answers is not in the document]
Comment (Algunenano) [reply src/Core/SettingsFields.cpp:263]: @groeneai Please handle this in this PR too

[the remark this answers is not in the document]
Comment (groeneai) [reply src/Core/SettingsFields.cpp:263]: Done in 6233c366.

`SettingFieldNonZeroUInt64::readBinary` now re-runs `checkValueNonZero()` after delegating to the base, mirroring the `parseFromString` override next to it. `read_binary` was the only one of the twelve `SettingFieldOps` entries that could store a forbidden value: `assign_from_field`, `parse_from_string`, `cast_value`, `value_to_string` and `string_to_value` all dispatch on the derived type, and `typed_copy` copies an already validated instance. `SettingFieldNonZeroUInt64` is also the only type in the tree that derives from a setting field, every other one is `final`, so there is no sibling site.

Two things I measured that shaped it, in case they matter to you.

**A query over TCP already rejected the zero before this commit**, so the regression cannot be a stateless test. `SettingsConstraints::checkOrClamp` casts every changed setting through `Settings::castValueUtil`, which builds a temporary of the derived type and assigns to it, so the check already ran there with the same `BAD_ARGUMENTS`. Frames 5 to 15 of the trace on a build without the fix:

```
SettingFieldNonZeroUInt64::operator=(Field const&)   <- SettingsFields.cpp:685
settingFieldCastValue<SettingFieldNonZeroUInt64> -> Settings::castValueUtil
getNewValueToCheck -> SettingsConstraints::checkImpl -> checkOrClamp
Context::checkSettingsConstraints  <- TCPHandler::processQuery:2819
```

The gap is real only where nothing runs after the decode: `QueryPlanSerializationSettings::readBinary`, which declares `grace_hash_join_initial_buckets`, `grace_hash_join_max_buckets` and `temporary_files_buffer_size` and has no constraints check anywhere in `src/Processors/`, and the `OLD_FORMAT` branch of `DistributedAsyncInsertHeader`. That is why the test drives `Settings::read` with `SettingsWriteFormat::BINARY` as a gtest instead of a `.sh`.

**One behaviour change I would rather you decided.** An old-format distributed batch written before `max_parallel_replicas` became `NonZeroUInt64` (dd6599868adb6cb, 2024-03) can carry an explicit `0`, and such a file would now stop draining, which `DistributedAsyncInsertHeader.cpp:155` says it specifically does not want. The mitigating measurement: the modern `STRINGS_WITH_FLAGS` header already rejects `0` for those settings today, through `parseFromString`, so this only makes the legacy branch behave like the mainstream one. If you would rather that branch normalize a zero instead of throwing, I will add it.

[the remark this answers is not in the document]
Comment (groeneai) [reply src/Core/SettingsFields.cpp:263]: Confirmed, and worse than I put it at 18:33. I had checked that the modern header path already rejects a zero, but not how the queue classifies the failure. `isDistributedSendBroken` (`DistributedAsyncInsertHelpers.cpp:39-50`) lists eight codes plus a local `ATTEMPT_TO_READ_AFTER_EOF`, and `BAD_ARGUMENTS` is not among them, so `processFile` rethrows without `markAsBroken` (`DistributedAsyncInsertDirectoryQueue.cpp:490`) and the file is retried indefinitely rather than quarantined. The wedge is real.

Two facts narrow it without refuting it. Nothing in the tree writes that signature any more (`DistributedSink.cpp:1053` writes only `DBMS_DISTRIBUTED_SIGNATURE_HEADER`), so the exposure is files already on disk from a version old enough to use `0xCAFECABE`. And the same input already wedges on master through the modern branch, which reads with `STRINGS_WITH_FLAGS` (`DistributedAsyncInsertHeader.cpp:67`) and so validates via `parseFromString`. My change widens that class rather than creating it.

One correction on the suggested remedy: grandfathering inside the `OLD_FORMAT` branch cannot be a `try` around the read. `insert_settings->read(in, BINARY)` consumes the settings straight from the file buffer with no length prefix and throws inside `accessor.readBinary` (`BaseSettings.h:800`), so `in` is left mid list and the `readStringBinary(insert_query, in)` on the next line would read the remaining setting bytes as the query text. Containing it needs a permissive flag threaded through `BaseSettings::read`, not a `catch` in the header.

The lever I would take instead is to drop the field override and validate in `QueryPlanSerializationSettings::readBinary`. Benefit: that is the only place the bypass has teeth. It declares `grace_hash_join_initial_buckets`, `grace_hash_join_max_buckets` and `temporary_files_buffer_size` as `NonZeroUInt64`, `checkSettingsConstraints` has zero occurrences under `src/Processors/`, and the one consumer-side guard, `clampTemporaryFilesBufferSize`, only bounds from above, so a peer-supplied zero reaches `SortingStep.cpp:472` as a temporary-file buffer size. Cost: `Settings` loses nothing, since the legacy TCP path is already covered by constraints as measured above, but re-running derived validation at collection level costs a `Field` round trip per changed setting per step on plan deserialization.

That asymmetry is why I put it on the field first and why I would now move it. I am happy to push either shape, including leaving it exactly as it is if you would rather keep the invariant on the field and accept the legacy-batch risk.

File: src/Core/SettingsFields.cpp [hunk 2]
```diff
@@ -679,6 +679,12 @@ void SettingFieldNonZeroUInt64::parseFromString(const String & str)
     checkValueNonZero();
 }
 
+void SettingFieldNonZeroUInt64::readBinary(ReadBuffer & in)
+{
+    SettingFieldUInt64::readBinary(in);
```
[the remark this answers is not in the document]
Comment (groeneai) [reply src/Core/SettingsFields.cpp:684]: Answered in the existing thread on line 263, where the same point was already open: https://github.com/ClickHouse/ClickHouse/pull/118610#discussion_r3951996634

Short version: the queue-classification half is confirmed, `BAD_ARGUMENTS` is absent from `isDistributedSendBroken` so the file is retried rather than quarantined. The first of the two suggested shapes is not available as written, because a `catch` around `insert_settings->read(in, BINARY)` leaves the file buffer mid list and the following `readStringBinary(insert_query, in)` would then read setting bytes as the query. The second shape is available and measured, and the choice between it and the current one is with the reviewer.

**Comment by groeneai** [comment]
### CI finish ledger - 6233c3669e3ad080fbfbfd5a73268291a9419fbe

Every failure below has an owner: a fixing PR (mine or external), or a full-effort fix task
whose fixing-PR link will be posted here when it opens. Only `CH Inc sync` is exempt.

| Check / test | Reason | Owner / fixing PR |
|---|---|---|
| Build profile diff | CI-infra: the `has_pr_data` probe raises `CI logs cluster query failed` on the `binary_sizes` read (`ci/jobs/build_profile_diff_job.py:721`) | #118685 (mine, open) |

Session id: cron:our-pr-ci-monitor:20260908-013052

**Review by Algunenano — APPROVED** [review]
LGTM
